### PR TITLE
[core] Fix internal storage S3 bugs

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

There are several S3 bugs in the internal storage. This PR fixes them. S3 tests are enabled in the unittest to ensure S3 is working properly with the fix.

1. `/bucket/foo` is an invalid S3 path, pyarrow would raise `E pyarrow.lib.ArrowInvalid: Missing bucket name in S3 URI` with any S3 path starts with `/`. However, if you does not start with `/`, the path will be resolved to `<local_workdir>/<s3 path>`, which is also wrong.
2. Miss `storage.delete_dir`. It become very hard to remove directories recursively.
3. When `storage.get/delete/delete_dir` is a missing S3 key/dir, it raises `OSError` instead of `FileNotFoundError` (`FileNotFoundError` is raised when using pyarrow.fs.LocalFilesystem). `FileNotFoundError` in `storage.get/delete/delete_dir` returns `None`, but `OSError` is unhandled.
4. When `storage.list` a S3 key, it does not raise `NotADirectoryError`
5. When under windows, "/" becomes "\\" and S3 does not support it.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/internal/storage.py [hunk 1]
```diff
@@ -129,6 +144,10 @@ def get(self, path: str) -> bytes:
                 return f.read()
         except FileNotFoundError:
             return None
+        except OSError as e:
+            if _is_os_error_file_not_found(e):
+                return None
```
Comment (iycheng) [inline python/ray/internal/storage.py:149]: How about this:

```
except Exception as e:
    if _is_file_not_found(e):
        return None
    raise e
```

File: python/ray/tests/mock_s3_server.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
+# extracted from aioboto3
```
Comment (iycheng) [inline python/ray/tests/mock_s3_server.py:1]: should we move the workflow one to here directly? (I'm ok with deleting that in another PR)

**Review by ericl — APPROVED** [review]

File: python/ray/tests/mock_s3_server.py (hunk 2 above)
Comment (suquark) [reply python/ray/tests/mock_s3_server.py:1]: yes, I will use this in workflow tests after this PR..

File: python/ray/internal/storage.py (hunk 1 above)
Comment (suquark) [reply python/ray/internal/storage.py:149]: I prefer to be more explicitly about the exception handling. According the current understanding, `OSError` is the extra exception we need to handle, and `_is_os_error_file_not_found` indicates this is a temporary patch. `Exception` lowers the level of exception handling too much.

**Comment by suquark** [comment]
There are some windows failures.

**Comment by suquark** [comment]
The CI errors are not related. I'll merge it.

**Pull request merged**